### PR TITLE
Add dedicated config file for cc local workers

### PR DESCRIPTION
**Pull request by iaftab-alam**
Before this PR, same config file was used for both ccng and the (two) cc local workers. This dedicated file aims to decouple this bond and provide a way to override the configurations for cc local workers.

Thanks for contributing to cloud_controller_ng. To speed up the process of reviewing your pull request please provide us with:

* Links to any other associated PRs

* [x] I have reviewed the [contributing guide](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/CONTRIBUTING.md)

* [x] I have viewed, signed, and submitted the Contributor License Agreement

* [x] I have made this pull request to the `main` branch

* [x] I have run all the unit tests using `bundle exec rake`

* [ ] I have run [CF Acceptance Tests](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/spec/README.md#cf-acceptance-tests-cats)

### CODE REVIEW

File: lib/tasks/rake_config.rb [hunk 1]
```diff
@@ -6,16 +6,30 @@ def context
 
     attr_writer :context
 
-    def config
+    def config(is_cc_local_worker: false)
```
Comment (kathap) [inline lib/tasks/rake_config.rb:9]: This function is used in many places in cc_ng.. 
if you want to access the config key is_cc_local_worker you could use RakeConfig.config.get(:key)

File: lib/cloud_controller/config.rb [hunk 2]
```diff
@@ -42,6 +34,22 @@ def schema_class_for_context(context, _config)
         const_get("VCAP::CloudController::ConfigSchemas::#{module_name}::#{context.to_s.camelize}Schema")
       end
 
+      def read_file(file_name)
+        deep_symbolize_keys_except_in_arrays(VCAP::CloudController::YAMLConfig.safe_load_file(file_name)) || {}
+      end
+
+      def load_from_hash(config_hash, context: :api, secrets_hash: {})
+        config = deep_symbolize_keys_except_in_arrays(config_hash)
```
Comment (philippthun) [inline lib/cloud_controller/config.rb:42]: You can remove this `deep_symbolize_keys_except_in_arrays` line from here as it is already done in `read_file`.

File: lib/cloud_controller/config.rb [hunk 3]
```diff
@@ -22,15 +22,7 @@ class InvalidConfigPath < StandardError
     class << self
       def load_from_file(file_name, context: :api, secrets_hash: {})
         config = VCAP::CloudController::YAMLConfig.safe_load_file(file_name)
```
Comment (philippthun) [inline lib/cloud_controller/config.rb:24]: You can replace this `safe_load_file` line with `read_file`.

So `load_from_file` does `read_file` + `load_from_hash`.

File: lib/tasks/rake_config.rb [hunk 4]
```diff
@@ -6,16 +6,30 @@ def context
...
       secrets_hash = {}
       # TODO: require secrets fetcher?
       secrets_hash = VCAP::CloudController::SecretsFetcher.fetch_secrets_from_file(secrets_file) unless secrets_file.nil?
 
-      VCAP::CloudController::Config.load_from_file(config_file, context:, secrets_hash:)
+      api_config = VCAP::CloudController::Config.load_from_file(config_file, context:, secrets_hash:)
+
+      if is_cc_local_worker
+        local_worker_config_hash = VCAP::CloudController::Config.read_file(cc_local_worker_config_file)
```
Comment (philippthun) [inline lib/tasks/rake_config.rb:17]: My proposal to minimize duplicate code execution is to:
1. read the ccng config file (`Config.read_file`)
2. if `context` is `:api` and local_worker_override file exists
  a. read the local_worker_override file (`Config.read_file`)
  b. merge config hashes
3. load config (`Config.load_from_hash`)

File: lib/tasks/rake_config.rb (hunk 1 above)
Comment (philippthun) [inline lib/tasks/rake_config.rb:9]: You don't need this `is_cc_local_worker` parameter but can use the `context` instead (that is either `:api` for local workers or `:worker` for generic workers.

File: config/cloud_controller_local_worker_override.yml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Configurations specific to cc local workers are overridden using this file in capi-release.
```
Comment (philippthun) [inline config/cloud_controller_local_worker_override.yml:1]: I don't think that you need this empty file. For testing purposes you can simply mock it.

File: lib/cloud_controller/config.rb (hunk 2 above)
Comment (iaftab-alam) [reply lib/cloud_controller/config.rb:42]: We need to keep this, first it is needed for deep merge with secret_hash as it expect a valid hash `config = config.deep_merge(secrets_hash)`, second, `deep_symbolize_keys_except_in_arrays` returns a hash if even passed an already hash.

File: config/cloud_controller_local_worker_override.yml (hunk 5 above)
Comment (iaftab-alam) [reply config/cloud_controller_local_worker_override.yml]: This is intended for development purposes only.

File: spec/unit/lib/cloud_controller/config_spec.rb [hunk 6]
```diff
@@ -48,6 +48,160 @@ module VCAP::CloudController
...
+        expect do
+          Config.read_file('nonexistent.yml')
+        end.to raise_error(Errno::ENOENT, /No such file or directory @ rb_sysopen - nonexistent.yml/)
+      end
+
+      context 'read a file' do
+        context 'when a file cannot be read' do
+          let(:file_name) { 'test_config.yml' }
+
+          it 'return an empty hash' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/config_spec.rb:62]:
```suggestion
          it 'returns an empty hash' do
```

File: spec/unit/lib/cloud_controller/config_spec.rb [hunk 7]
```diff
@@ -48,6 +48,160 @@ module VCAP::CloudController
...
+          end
+
+          let(:cc_local_worker_config_file) do
+            file = Tempfile.new('cc_local_config_file.yml')
+            file.write(YAML.dump(config_contents))
+            file.close
+            file
+          end
+
+          it 'return a valid hash' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/config_spec.rb:86]:
```suggestion
          it 'returns a valid hash' do
```

File: spec/unit/lib/cloud_controller/config_spec.rb [hunk 8]
```diff
@@ -48,6 +48,160 @@ module VCAP::CloudController
...
+        end
+
+        context 'when empty YAML file is provided' do
+          let(:cc_local_worker_config_file) do
+            config = YAMLConfig.safe_load_file('config/cloud_controller_local_worker_override.yml')
+            file = Tempfile.new('cc_local_worker_config.yml')
+            file.write(YAML.dump(config))
+            file.close
+            file
+          end
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/config_spec.rb:99]:
```suggestion
          let(:cc_local_worker_config_file) { 'config/cloud_controller_local_worker_override.yml' }
```

File: spec/unit/lib/cloud_controller/config_spec.rb [hunk 9]
```diff
@@ -48,6 +48,160 @@ module VCAP::CloudController
...
+
+          it 'return an empty hash' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/config_spec.rb:101]:
```suggestion
          it 'returns an empty hash' do
```

File: spec/unit/lib/cloud_controller/config_spec.rb [hunk 10]
```diff
@@ -48,6 +48,160 @@ module VCAP::CloudController
...
+    end
+
+    describe '.load_from_hash' do
+      it 'raises error if an empty hash is provided' do
+        expect do
+          Config.load_from_hash({}, context: :api)
+        end.to raise_error(Membrane::SchemaValidationError)
+      end
+
+      context 'merges default values' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/config_spec.rb:116]: I would remove this outer context and move `context 'when no config values are provided'` one level up.

File: spec/unit/lib/cloud_controller/config_spec.rb [hunk 11]
```diff
@@ -48,6 +48,160 @@ module VCAP::CloudController
...
+
+          it 'has the default values' do
+            expect(config_hash[:db][:max_connections]).to eq(42)
+            expect(config_hash[:name]).to eq('api')
+            expect(config_hash[:local_route]).to eq('127.0.0.1')
+          end
+        end
+      end
+
+      context 'unit test' do
```
Comment (philippthun) [inline spec/unit/lib/cloud_controller/config_spec.rb:134]: The tests in this context are redundant with what is in `describe '.load_from_file'`. If you want to keep them here, you should remove them from below. Then also `context 'unit test'` should be changed to something else...

File: spec/unit/lib/cloud_controller/config_spec.rb [hunk 12]
```diff
@@ -48,6 +48,160 @@ module VCAP::CloudController
...
 
+    describe '.read_file' do
+      it 'raises error if the file does not exist' do
+        expect do
+          Config.read_file('nonexistent.yml')
+        end.to raise_error(Errno::ENOENT, /No such file or directory @ rb_sysopen - nonexistent.yml/)
+      end
+
+      context 'read a file' do
+        context 'when a file cannot be read' do
```
[the remark this answers is not in the document]
Comment (iaftab-alam) [reply spec/unit/lib/cloud_controller/config_spec.rb:59]: right, missed that in the same context, for empty file there is a unit test...

**Review by philippthun — APPROVED** [review]

**Pull request merged**